libmscore/changeMap.cpp

```diff
@@ -0,0 +1,487 @@
+            }
+      }
+
+//---------------------------------------------------------
+//   addFixed
+//---------------------------------------------------------
+
+void ChangeMap::addFixed(Fraction tick, int value)
+      {
+      insert(tick, ChangeEvent(tick, value));
+      cleanedUp = false;
+      }
+
+//---------------------------------------------------------
+//   addRamp
+///   A `change` of 0 means that the change in velocity should be calculated from the next fixed
+///   velocity event.
+//---------------------------------------------------------
+
+void ChangeMap::addRamp(Fraction stick, Fraction etick, int change, ChangeMethod method, ChangeDirection direction)
+      {
+      change = abs(change);
+      change *= (direction == ChangeDirection::INCREASING) ? 1 : -1;
+      insert(stick, ChangeEvent(stick, etick, change, method, direction));
+      cleanedUp = false;
+      }
+
+//---------------------------------------------------------
+//   cleanup
+//---------------------------------------------------------
+
+void ChangeMap::cleanup()
+      {
+      //
+      // Stage 0: put the ramps in size order if they start at the same point
+      //
+      for (auto& tick : uniqueKeys()) {
+            // `events` will contain all the ramps at this tick
+            std::vector<ChangeEvent> events;
+            for (auto& event : values(tick)) {
+                  if (event.type == ChangeEventType::FIX)
+                        continue;
+                  events.push_back(event);
+                  }
+
+            if (int(events.size()) > 1) {
+                  // new order is a map of the NEGATIVE length of a ramp to a copy of its event
+                  QMultiMap<Fraction, ChangeEvent> newOrder;
+                  for (auto& event : events) {
+                        newOrder.insert(event.tick - event.etick, event);
+                        // And clear out the events at this tick
+                        remove(tick, event);
+                        }
+
+                  // newOrder's values are implicitally ordered by the key, which is the
+                  // NEGATIVE length of the ramp.
+                  for (auto& event : newOrder.values()) {
+                        insert(tick, event);
+                        }
+                  }
+            }
+
+      //
+      // Stage 1: remove any ramps or fixes that are completely enclosed within other ramps
+      //
+      Fraction currentRampStart = Fraction(-1, 1);    // start point of ramp we're in
+      Fraction currentRampEnd = Fraction(-1, 1);      // end point of ramp we're in
+      Fraction lastFix = Fraction(-1, 1);             // the position of the last fix event
+      bool inRamp = false;                            // whether we're in a ramp or not
+
+      // Keep a record of the endpoints
+      std::vector<std::pair<Fraction, Fraction>> endPoints;
+      std::vector<bool> startsInRamp;
+
+      for (ChangeEvent& event : values()) {
```

Comment:
Continuing my statements about `tick` variable, you could use STL-style iterators and deal with tick-event pairs in this loop so no tick member variable is needed.

Reply:
Yeah, I guess - it was more a design decision because I'm not a massive fan of working with STL-iterators, but I'm sure it won't be too much trouble to change it to that, if you think it would be better :)

Reply:
My main point here is not about working with iterators but rather about not binding change events data to absolute tick values, using iterators here is just a way to obtain tick values directly from the map.